GithubBrowserSample/app/src/main/java/com/android/example/github/util/AutoClearedValue.kt

```diff
@@ -16,26 +16,31 @@
 
 package com.android.example.github.util
 
-import androidx.lifecycle.Lifecycle
-import androidx.lifecycle.LifecycleObserver
-import androidx.lifecycle.OnLifecycleEvent
 import androidx.fragment.app.Fragment
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.observe
 import kotlin.properties.ReadWriteProperty
 import kotlin.reflect.KProperty
 
 /**
- * A lazy property that gets cleaned up when the fragment is destroyed.
+ * A lazy property that gets cleaned up when the fragment's view is destroyed.
  *
- * Accessing this variable in a destroyed fragment will throw NPE.
+ * Accessing this variable while the fragment's view is destroyed will throw NPE.
  */
 class AutoClearedValue<T : Any>(val fragment: Fragment) : ReadWriteProperty<Fragment, T> {
     private var _value: T? = null
 
     init {
-        fragment.lifecycle.addObserver(object : LifecycleObserver {
-            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
-            fun onDestroy() {
-                _value = null
+        fragment.lifecycle.addObserver(object: DefaultLifecycleObserver {
+            override fun onCreate(owner: LifecycleOwner) {
+                fragment.viewLifecycleOwnerLiveData.observe(fragment) { viewLifecycleOwner ->
+                    viewLifecycleOwner?.lifecycle?.addObserver(object: DefaultLifecycleObserver {
+                        override fun onDestroy(owner: LifecycleOwner) {
+                            _value = null
+                        }
+                    })
+                }
```

Comment:
Hi @ianhanniballake, I am trying to understand why we observe `viewLifecycleOwnerLiveData` in `onCreate` of Fragment, I tried removing the `fragment.lifecycle` observer and observe for `fragment.viewLifecycleOwnerLiveData` right away, seems working fine, I would really appreciate if you will explain this. And also should we consider remove observer `onDestroy` ?

Reply:
There's no requirement that you create a Fragment instance on the main thread - if you try to directly observe the `fragment.viewLifecycleOwnerLiveData` when on a different thread, you'll hit an exception. By waiting for `onCreate()`, we ensure that we are always calling `observe()` on the main thread (and always in the same relative order compared to other observers).

There's no reason to explicitly remove an observer in `onDestroy()` as all references to that Lifecycle are dropped immediately after the Lifecycle is destroyed. In addition, `onDestroy()` is a terminal state meaning any Observers associated with it will never get any additional callbacks.

Reply:
Thank you very much for your quick response.